Parallelization IO test fails due to keeping the Class level variable use_dll = True

**Error:**

The test `test_parallelizaion_io` passes when being executed as a single test with `pytest`.

When running all tests with consecutively `test_parallelizaion_io` fails with

```
>       write_solution_times = sim.root.input['return'].get('write_solution_times', 0)
E       KeyError: 'return'

cadet\cadet_dll.py:1831: KeyError
```


**Cause:**
When `pytest` is called, all tests use the same class instances. In the earlier executed `test_dll` the `install_path` is set to `<something>.dll` which causes the `CadetMeta`class to set `use_dll = True`  on a class level, which forces "dll-mode" for all tests coming afterwards. 

Thus, when calling `test_parallelization_io` and executing

```    
model1 = Cadet()
[...]
data = model.run_load()
```

it attempts to run the model in "dll-mode" which fails since `run_load` requires a "return" key there.

**Underlying issue:**
In the `Cadet()` class `use_dll` is defined as an instance variable but when provided an 'install_path' to a .dll will be overwritten with `use_dll = True` from `CadetMeta` with `cls.use_dll = cadet_path.suffix in [".dll", ".so"]` in `cadet.py` line 156.


**Solution:**
I am not super familiar with the underlying mechanism of `CadetMeta` and why it exists in a way, where it can overwrite/assume variables which are not explicitly set. However,  I created a function Cadet.reset() that resets the cadet paths and added it to a fixture that automatically runs after every test and thus exits the "dll-mode" forced by the previously defined Metclass. Maybe we should see if there is a more elegant fix in the future, but with this, all tests run without errors now.

--- a/cadet/cadet.py
+++ b/cadet/cadet.py
@@ -523,6 +523,13 @@ class Cadet(H5, metaclass=CadetMeta):
         if runner is not None:
             runner.clear()
 
+    def delete_file(self) -> None:
+        if self.filename is not None:
+            try:
+                os.remove(self.filename)
+            except FileNotFoundError:
+                pass
+
     def __del__(self):
         self.clear()
         del self._cadet_dll_runner

--- a/cadet/cadet.py
+++ b/cadet/cadet.py
@@ -213,6 +213,9 @@ class Cadet(H5, metaclass=CadetMeta):
         self.cadet_create_lwe_path: Optional[Path] = None
         self.return_information: Optional[dict] = None
 
+        self._cadet_cli_runner: Optional[CadetCLIRunner] = None
+        self._cadet_dll_runner: Optional[CadetDLLRunner] = None
+
         # Regardless of settings in the Meta Class, if we get an install_path, we respect the install_path
         if install_path is not None:
             self.use_dll = use_dll

--- a/cadet/cadet.py
+++ b/cadet/cadet.py
@@ -112,6 +112,10 @@ class CadetMeta(type):
     This meta class allows setting the `cadet_path` attribute for all instances of the
     `Cadet` class.
     """
+    use_dll = False
+    cadet_cli_path = None
+    cadet_dll_path = None
+    cadet_create_lwe_path = None
 
     @property
     def cadet_path(cls) -> Optional[Path]:
@@ -131,7 +135,7 @@ class CadetMeta(type):
             return None
 
     @cadet_path.setter
-    def cadet_path(cls, cadet_path: os.PathLike) -> None:
+    def cadet_path(cls, cadet_path: Optional[os.PathLike]) -> None:
         """
         Set the CADET path and initialize the appropriate runner.
 
@@ -145,13 +149,20 @@ class CadetMeta(type):
         If the path is a DLL, a `CadetDLLRunner` runner is used.
         Otherwise, a `CadetFileRunner` runner is used.
         """
-        cadet_path = Path(cadet_path)
-
         warnings.warn(
-            "Support for setting cadet.cadet_path will be removed in a future version. "
-            "TODO: What alternative should be used?",
+            "Support for setting Cadet.cadet_path will be removed in a future version. "
+            "Please set the `install_path` on instance level.",
             DeprecationWarning
         )
+        if cadet_path is None:
+            cls.use_dll = False
+            cls._install_path = None
+            cls.cadet_cli_path = None
+            cls.cadet_dll_path = None
+            cls.cadet_create_lwe_path = None
+            return
+
+        cadet_path = Path(cadet_path)
 
         cls.use_dll = cadet_path.suffix in [".dll", ".so"]
 
@@ -314,9 +325,9 @@ class Cadet(H5, metaclass=CadetMeta):
         Otherwise, a `CadetFileRunner` runner is used.
         """
         cadet_path = Path(cadet_path)
-
         warnings.warn(
-            "Deprecation warning: Support for setting cadet.cadet_path will be removed in a future version.",
+            "Deprecation warning: Support for setting cadet.cadet_path will be removed "
+            " in a future version. Use `install_path` instead.",
             FutureWarning
         )
         self.install_path = cadet_path

--- a/tests/test_autodetection.py
+++ b/tests/test_autodetection.py
@@ -1,6 +1,0 @@
-from cadet import Cadet
-
-
-def test_autodetection():
-    sim = Cadet()
-    assert sim.cadet_runner is not None

--- a/tests/test_install_path_settings.py
+++ b/tests/test_install_path_settings.py
@@ -47,7 +47,7 @@ def test_install_path():
 def test_dll_runner_attrs():
     cadet = Cadet(full_path_dll.parent.parent)
     cadet_runner = cadet._cadet_dll_runner
-    assert re.match("\d\.\d\.\d", cadet_runner.cadet_version)
+    assert re.match(r"\d\.\d\.\d", cadet_runner.cadet_version)
     assert isinstance(cadet_runner.cadet_branch, str)
     assert isinstance(cadet_runner.cadet_build_type, str | None)
     assert isinstance(cadet_runner.cadet_commit_hash, str)
@@ -58,7 +58,7 @@ def test_dll_runner_attrs():
 def test_cli_runner_attrs():
     cadet = Cadet(full_path_dll.parent.parent)
     cadet_runner = cadet._cadet_cli_runner
-    assert re.match("\d\.\d\.\d", cadet_runner.cadet_version)
+    assert re.match(r"\d\.\d\.\d", cadet_runner.cadet_version)
     assert isinstance(cadet_runner.cadet_branch, str)
     assert isinstance(cadet_runner.cadet_build_type, str | None)
     assert isinstance(cadet_runner.cadet_commit_hash, str)

--- a/tests/test_meta_class_install_path.py
+++ b/tests/test_meta_class_install_path.py
@@ -22,7 +22,7 @@ def test_meta_class():
     assert sim.use_dll
     assert sim.install_path == full_path_dll.parent.parent
     assert sim.cadet_dll_path == full_path_dll
-    assert sim.cadet_cli_path.parent == full_path_dll.parent
+    assert sim.cadet_cli_path.parent.parent == full_path_dll.parent.parent
 
     # With an install path given, the sim instance should use the given install path
     sim = Cadet(install_path=install_path_conda)

--- a/tests/test_meta_class_install_path.py
+++ b/tests/test_meta_class_install_path.py
@@ -29,3 +29,10 @@ def test_meta_class():
     assert sim.install_path == install_path_conda
     assert sim.cadet_dll_path.parent.parent == install_path_conda
     assert sim.cadet_cli_path.parent.parent == install_path_conda
+
+    # Reset Path
+    Cadet.cadet_path = None
+
+
+if __name__ == "__main__":
+    pytest.main([__file__])

--- a/tests/test_parallelization.py
+++ b/tests/test_parallelization.py
@@ -8,6 +8,8 @@ n_jobs = 2
 def run_simulation(model):
     model.save()
     data = model.run_load()
+    model.delete_file()
+
     return data
 
 

--- a/tests/test_parallelization.py
+++ b/tests/test_parallelization.py
@@ -40,3 +40,7 @@ def test_parallelization_simulation():
         delayed(run_simulation)(model, ) for model in models
     )
     assert results_sequential == results_parallel
+
+
+if __name__ == "__main__":
+    pytest.main([__file__])

